hadoop-hdfs-project/hadoop-hdfs/src/main/java/org/apache/hadoop/ozone/container/common/states/datanode/RunningDatanodeState.java

```diff
@@ -0,0 +1,292 @@
+      createNewContainerID(Path idPath)
+      throws IOException {
+    StorageContainerDatanodeProtocolProtos.ContainerNodeIDProto
+        containerIDProto = StorageContainerDatanodeProtocolProtos
+        .ContainerNodeIDProto.newBuilder()
+        .setDatanodeID(createDatanodeID().getProtoBufMessage()).build();
+    try (FileOutputStream stream = new FileOutputStream(idPath.toFile())) {
+      stream.write(containerIDProto.toByteArray());
+      return containerIDProto;
+    }
+  }
+
+  /**
+   * Returns ContainerNodeIDProto or null in case of Error.
+   *
+   * @return ContainerNodeIDProto
+   */
+  private StorageContainerDatanodeProtocolProtos.ContainerNodeIDProto
+      getContainerNodeID() {
+    String dataNodeIDPath = conf.get(OzoneConfigKeys.OZONE_SCM_DATANODE_ID);
+    if (dataNodeIDPath == null || dataNodeIDPath.isEmpty()) {
+      LOG.error("A valid file path is needed for config setting {}",
+          OzoneConfigKeys.OZONE_SCM_DATANODE_ID);
+
+      // This is an unrecoverable error.
+      this.context.setState(DatanodeStateMachine.DatanodeStates.SHUTDOWN);
+    }
+    StorageContainerDatanodeProtocolProtos.ContainerNodeIDProto nodeID;
+    // try to read an existing ContainerNode ID.
+    try {
+      nodeID = readPersistedDatanodeID(Paths.get(dataNodeIDPath));
+      if (nodeID != null) {
+        return nodeID;
+      }
+    } catch (IOException ex) {
+      LOG.trace("Not able to find container Node ID, creating it.", ex);
+    }
+    // Not found, let us create a new datanode ID, persist it and return that
+    // info to SCM.
+    try {
+      nodeID = createNewContainerID(Paths.get(dataNodeIDPath));
+      return nodeID;
+    } catch (IOException ex) {
+      LOG.error("Creating new node ID failed.", ex);
+    }
+    return null;
+  }
+
+  /**
+   * Called before entering this state.
+   */
+  @Override
+  public void onEnter() {
+    LOG.trace("Entering handshake task.");
+  }
+
+  /**
+   * Called After exiting this state.
+   */
+  @Override
+  public void onExit() {
+    LOG.trace("Exiting handshake task.");
+  }
+
+  /**
+   * Executes one or more tasks that is needed by this state.
+   *
+   * @param executor -  ExecutorService
+   */
+  @Override
+  public void execute(ExecutorService executor) {
+    ecs = new ExecutorCompletionService<>(executor);
+    for (EndpointStateMachine endpoint : connectionManager.getValues()) {
+      Callable<EndpointStateMachine.EndPointStates> endpointTask
+          = getEndPointTask(endpoint);
+      ecs.submit(endpointTask);
+    }
+  }
+
+  private Callable<EndpointStateMachine.EndPointStates>
+      getEndPointTask(EndpointStateMachine endpoint) {
+    switch (endpoint.getState()) {
+    case GETVERSION:
+      return new VersionEndpointTask(endpoint, conf);
+    case REGISTER:
+      RegisterEndpointTask task = new RegisterEndpointTask(endpoint, conf);
+      task.setContainerNodeIDProto(getContainerNodeID());
```

Comment:
NIT: maybe use a builder pattern to avoid creating RegisterEndpointTask without ContainerNodeID.